src/memory.c

```diff
@@ -109,3 +116,196 @@ void memory_free_page(void *pageaddr) {
     freemap[cellnumber] |= cellmask;
     pages_free++;
 }
+
+struct kmalloc_page_info * kmalloc_create_page_info(unsigned paddr) {
+    struct kmalloc_page_info *n = (struct kmalloc_page_info *)paddr;
+    n->max_free_gap = KMALLOC_NUM_SLOTS;
+    memset(n->free, 1, KMALLOC_SLOT_SIZE * KMALLOC_NUM_SLOTS);
+    return (struct kmalloc_page_info *)n;
+}
+
+void kmalloc_get_page() {
+    //get current process's pagetable
+    struct pagetable *pt = current->pagetable;
+    unsigned vaddr = kmalloc_next_vaddr;
+    unsigned phys_addr;
+
+    if (kmalloc_next_vaddr >= PROCESS_ENTRY_POINT) {
+        console_printf("kmalloc is out of space... uh oh\n");
+        //TODO: figure out a system interrupt of some sort here
+        return;
+    }
+ 
+    if (pagetable_getmap(pt, vaddr, &phys_addr) != 1) {
+        console_printf("kmalloc: Unable to get mapping to next vmem page\n");
+        return;
+    }
+
+    console_printf("Got page with vaddr: %x, and paddr: %x\n", vaddr, phys_addr);
+
+    //prepare for the next page requested by updating the global variable for it.
+    kmalloc_next_vaddr += PAGE_SIZE;
+    
+    // build the page struct
+    struct kmalloc_page_info *pg_info = kmalloc_create_page_info(phys_addr);
+    
+ 
+    //rearrange the linked list of pages with the new page at the beginning.
+    pg_info->next = kmalloc_head;
+    kmalloc_head = pg_info;
+}
+
+int kmalloc_locate_sufficient_gap(struct kmalloc_page_info *page_info, int num_slots) {
```

Comment:
@kkoser Do these follow the Doxygen standard we set out for the .h files? Or is it just a multiline comment that describes how it helps.